From 495f95bd63a2e792c7bcccffb7f00ccaaafc7249 Mon Sep 17 00:00:00 2001
From: James Hogan <james.hogan@imgtec.com>
Date: Wed, 17 Jun 2015 14:55:32 +0100
Subject: [PATCH] MIPS: Disable s/w prefetch on interAptiv

The prefetch macros may be called with NULL pointers, and that is
ordinarily fine. However with 2G EVA mode accessing virtual address 0x0
causes a bus error. Prefetches don't normally cause bus errors on
interAptiv, but there is an issue, still under investigation, where
prefetches on one VPE which would trigger bus errors if they were normal
accesses can occasionally result in a bus error on the other VPE,
depending on what code the other VPE is executing.

Workaround these bus errors by removing software prefetch instructions
if CONFIG_EVA_2G is enabled.

Signed-off-by: James Hogan <james.hogan@imgtec.com>
---
 arch/mips/include/asm/processor.h | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/arch/mips/include/asm/processor.h b/arch/mips/include/asm/processor.h
index 1470b7b68b0e..1b608cfe4d25 100644
--- a/arch/mips/include/asm/processor.h
+++ b/arch/mips/include/asm/processor.h
@@ -344,12 +344,34 @@ unsigned long get_wchan(struct task_struct *p);
 
 #ifdef CONFIG_CPU_HAS_PREFETCH
 
+#ifdef CONFIG_EVA_2GB
+
+/*
+ * On interAptiv, a prefetch to an address which would ordinarily cause a bus
+ * error can under certain circumstances cause a different VPE to get a bus
+ * error. Lets avoid that!
+ */
+
+#define ARCH_HAS_PREFETCH
+static inline void prefetch(const void *ptr)
+{
+}
+
+#define ARCH_HAS_PREFETCHW
+static inline void prefetchw(const void *ptr)
+{
+}
+
+#else
+
 #define ARCH_HAS_PREFETCH
 #define prefetch(x) __builtin_prefetch((x), 0, 1)
 
 #define ARCH_HAS_PREFETCHW
 #define prefetchw(x) __builtin_prefetch((x), 1, 1)
 
+#endif
+
 /*
  * See Documentation/scheduler/sched-arch.txt; prevents deadlock on SMP
  * systems.
-- 
2.3.6

